### PR TITLE
Fix `CooldownMapping` generic typing and ensure other public methods/types have proper generics

### PR DESCRIPTION
**Pull request by bryanforbes**
## Summary

The title says it all

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/ext/commands/cooldowns.py [hunk 1]
```diff
@@ -50,7 +49,7 @@
     'MaxConcurrency',
 )
 
-T = TypeVar('T')
+T_contra = TypeVar('T_contra')
```
Comment (Gobot1234) [inline discord/ext/commands/cooldowns.py:52]: Seems like you're missing contravariant=True here

Comment (bryanforbes) [reply discord/ext/commands/cooldowns.py:52]: Ugh, I took it out to test something... _sigh_

**Comment by Rapptz** [comment]
Why make this contravariant?

**Comment by bryanforbes** [comment]
> Why make this contravariant?

If the type is only being used in a parameter, it can be either invariant or contravariant. Since `commands.Command._bucket` is typed as `CooldownMapping[Context[Any]]` (so `get_bucket()` will accept a `Context` or subclass), an invariant generic for `CooldownMapping` technically disallows assigning the mapping from `CooldownMapping.from_cooldown()` with a `BucketType` because that results in `CooldownMapping[Message | Context[Any]]`. Contravariance is the solution to that problem.

**Pull request merged**